File: Jiho/week3/day13/섬 연결하기.js
function solution(n, costs) {
  let answer = 0;
  const parent = Array(n)
    .fill(0)
    .map((_, index) => index);

  costs.sort((a, b) => a[2] - b[2]);

  function find(node) {
    if (parent[node] === node) return node;
    return (parent[node] = find(parent[node]));
  }

  function union(u, v) {
    const root_u = find(u);
    const root_v = find(v);
    if (root_u !== root_v) parent[root_v] = root_u;
  }

  let edgeCount = 0;

  for (const [island1, island2, cost] of costs) {
    if (find(island1) !== find(island2)) {
      union(island1, island2);
      answer += cost;
      edgeCount++;
    }

    if (edgeCount === n - 1) break;
  }

  return answer;
}

File: Jin/week03/day13/섬연결하기.java
package Jin.week03.day13;

import java.util.*;

class 섬연결하기 {
    static int[] parent;
    public int solution(int n, int[][] costs) {
        Arrays.sort(costs, Comparator.comparingInt(o -> o[2]));

        // 초기화
        parent = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }

        int answer = 0;
        int edge = 0;

        for (int i = 0; i < costs.length; i++) {
            // 정점 N개에서 신장 트리의 간선은 n - 1개이니까!
            if (edge == n - 1) break;

            if (find(costs[i][0]) != find(costs[i][1])) {
                union(costs[i][0], costs[i][1]);
                answer += costs[i][2];
                edge++;
            }
        }
        return answer;
    }

    private static int find(int x) {
        if (parent[x] == x) return x;
        return parent[x] = find(parent[x]);
    }

    private static void union(int x, int y) {
        int root1 = find(x);
        int root2 = find(y);
        parent[root2] = root1;
    }
}